### ISSUE TITLE
[Proposal]: Allow member names to match their containing type name

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/csharplang#5758]
# Allow member names to match their containing type name

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Removes the restriction where the compiler produces an error if a member name matches its containing type name:

> error CS0542: '{0}': member names cannot be the same as their enclosing type

## Motivation
[motivation]: #motivation

There are use cases where a developer will want a member name to match its containing type name.

## Detailed design
[design]: #detailed-design

It's possible that we'll need to only remove this code:

https://github.com/dotnet/roslyn/blob/5193908c7283d377b9133a24e0e4394c603869c6/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L1272-L1290

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by Youssef1313** [comment]
I don't know if this was the reason for the error, but I don't see this being confusing. It is not hard for beginners to spot an extra `void` (or return type) in my opinion.

**Comment by YairHalberstadt** [comment]
> It's possible that we'll need to only remove this code:

> https://github.com/dotnet/roslyn/blob/5193908c7283d377b9133a24e0e4394c603869c6/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L1272-L1290

I don't think it would be that simple, since you would want to disallow this for a nested type, which would almost always be ambiguous whether you're referring to the inner or outer type. There's probably other cases I'm not thinking of

**Comment by 333fred** [comment]
Unless there is an ldm champion, this needs to start as a discussion, not an issue. As such, I'm converting it over.
